Use origin/branch when checking for new commits + python rewrite

thisisrandy opened the pull request:
This fixes /issues/14 for me (I have coc.nvim checked out from the release branch).

Additionally, there was some other race-condition-like behavior going on where the on_stdout callback was apparently entered variably fewer times than expected (see 72c9dcf for details). I was unable to resolve it in vimscript, but a python rewrite did the trick.

thisisrandy committed:
Use branch when checking for new commits

thisisrandy committed:
- Added origin/[branch] to rev-list call
- Check v:vim_did_enter and call for debugging convenience

thisisrandy committed:
Added wait to the end of remote update chain

thisisrandy committed:
No need to say len(keys(...)) (this matches other branches better)

thisisrandy committed:
Rewrite in python

There was a lot of strange behavior in vimscript version of this plugin.
The on_stdout callback appeared to be called a variable number of times
before the on_exit callback, most likely indicating a race. I tried
refactoring to call the rev-list command sequentially for each plugin in
an async cascade, but for unknown reasons, stdout was blank for about
half of the calls in my setup.

In python, we can use a comparatively simple synchronous function
wrapped in vim.async_call, so that's what I've done. The plugin now
appears to finally work consistently and as expected. Since
async/job.vim is no longer needed, I've removed that as well.

thisisrandy committed:
Properly casts result of vim.eval to int

vim.eval returns a string, whereas we need an int for boolean testing

thisisrandy committed:
Adds g:outdated_plugins_trigger_mode

thisisrandy commented:
Oh, I also added a new option g:outdated_plugins_trigger_mode to trigger :PlugUpdate as needed

thisisrandy committed:
Update to for python3

Was using python2 subprocess.Popen, which accepts the text argument.
Since python3 is utf8, this arg was removed (always returns utf8).
Also got rid of unused stderr var to quiet lint

thisisrandy committed:
Decode process output from binary and use run()

Got it backwards in the previous commit. subprocess always returns
binary output now, so it must be decoded.

Also used subprocess.run instead of Popen/communicate for simplicity

thisisrandy committed:
Add guard to keep plug.vim from throwing exception

thisisrandy committed:
Add .gitignore

semanser commented:
hi @thisisrandy! Thanks for your work and sorry for my late reply. 

I've tried it and it works much better than vimscript version. Good job!

What is the minimal version of python required? I want to update the README file.

semanser commented:
It seems like I'm not able to add commits to this PR (I can't push to master branch on your forked repo).

thisisrandy commented:
Thanks for your reply, @semanser. I've added you as a collaborator on my fork, so you should have push access now. I developed the python bit using 3.6.8, so I know that's safe. Probably should have added the future imports to make it backward compatible, though. I'll take care of that in a bit, or if you beat me to it, you're welcome to do the same.

thisisrandy commented:
Actually, I take that back. The `subprocess` API changed in python 3, and it seems the `run` method, which I used, was introduced in 3.5. So 3.5 is the minimum requirement.

sitiom commented:
Just a comment here. Due to the python rewrite, when used along with [vimsence](https://github.com/hugolgst/vimsence), I get a freeze in startup until the plugin message appears. Depending on my connectivity, the freeze would take from noticeable to a complete freeze.

thisisrandy committed:
Change installation instructions to fork path

Note: this commit should be excluded from any pull requests, but it is
useful for clarity should anyone want to use the fork directly

thisisrandy committed:
Formatting

thisisrandy committed:
Use master if plug["branch"] is empty

It's unclear when this broke, but g:plugs[name].branch is apparently
now empty when not explicitly specified. Since this plugin never would
have worked if it was always that way, presumably the behavior changed
at some point

thisisrandy committed:
Add .venv + generalize pycache in gitignore

thisisrandy committed:
Add poetry

thisisrandy committed:
Rewrite as remote plugin to avoid blocking

thisisrandy committed:
Fix grammar

thisisrandy committed:
Add autocmd to call plugin fn to config section

thisisrandy committed:
Clarify wording/grammar + addt. info

thisisrandy committed:
Add note about trigger mode default value

thisisrandy committed:
Specify that the plugin is non-blocking

thisisrandy committed:
Specify why the plugin is needed alongside vim-plug

thisisrandy committed:
Add note about trigger mode being disruptive

thisisrandy committed:
Move autocmd to install section

thisisrandy committed:
Move requirements to their own section

thisisrandy committed:
Update badges

thisisrandy committed:
Add development section

thisisrandy committed:
Don't guess branch if not specified explicitly

thisisrandy committed:
Merge pull request #2 from thisisrandy/bad-branch

Don't guess branch if not specified explicitly

thisisrandy committed:
Use echom instead of echo to print messages
